src/core/api/notification_policy.go

```diff
@@ -0,0 +1,409 @@
+// displays event type, status(enabled/disabled), create time, last trigger time
+func (w *NotificationPolicyAPI) ListGroupByEventType() {
+	projectID := w.project.ProjectID
+	if !w.validateRBAC(rbac.ActionList, projectID) {
+		return
+	}
+
+	res, err := notification.PolicyMgr.List(projectID)
+	if err != nil {
+		w.SendInternalServerError(fmt.Errorf("failed to list notification policies by projectID %d: %v", projectID, err))
+		return
+	}
+
+	policies, err := constructPolicyWithTriggerTime(res)
+	if err != nil {
+		w.SendInternalServerError(fmt.Errorf("failed to list the notification policy trigger information: %v", err))
+		return
+	}
+	w.WriteJSONData(policies)
+}
+
+// Delete ...
+func (w *NotificationPolicyAPI) Delete() {
+	projectID := w.project.ProjectID
+	if !w.validateRBAC(rbac.ActionDelete, projectID) {
+		return
+	}
+
+	id, err := w.GetIDFromURL()
+	if id < 0 || err != nil {
+		w.SendBadRequestError(errors.New("invalid notification policy ID"))
+		return
+	}
+
+	policy, err := notification.PolicyMgr.Get(id)
+	if err != nil {
+		w.SendInternalServerError(fmt.Errorf("failed to get the notification policy %d: %v", id, err))
+		return
+	}
+	if policy == nil {
+		w.SendNotFoundError(fmt.Errorf("notification policy %d not found", id))
+		return
+	}
+
+	if policy.ProjectID == 0 {
+		w.SendNotFoundError(fmt.Errorf("notification policy %d with projectID %d not found", id, projectID))
+		return
+	}
+
+	if projectID != policy.ProjectID {
+		w.SendBadRequestError(fmt.Errorf("notification policy %d with projectID %d not belong to project %d in URL", id, policy.ProjectID, projectID))
+		return
+	}
+
+	if err = notification.PolicyMgr.Delete(id); err != nil {
+		w.SendInternalServerError(fmt.Errorf("failed to delete notification policy %d: %v", id, err))
+		return
+	}
+}
+
+// Test ...
+func (w *NotificationPolicyAPI) Test() {
+	projectID := w.project.ProjectID
+	if !w.validateRBAC(rbac.ActionCreate, projectID) {
+		return
+	}
+
+	policy := &models.NotificationPolicy{}
+	isValid, err := w.DecodeJSONReqAndValidate(policy)
+	if !isValid {
+		w.SendBadRequestError(err)
+		return
+	}
+
+	if !w.validateTargets(policy) {
+		return
+	}
+
+	if err := notification.PolicyMgr.Test(policy); err != nil {
+		w.SendBadRequestError(fmt.Errorf("notification policy %s test failed: %v", policy.Name, err))
+		return
+	}
+}
+
+func (w *NotificationPolicyAPI) validateRBAC(action rbac.Action, projectID int64) bool {
+	if w.SecurityCtx.IsSysAdmin() {
+		return true
+	}
+
+	project, err := w.ProjectMgr.Get(projectID)
+	if err != nil {
+		w.ParseAndHandleError(fmt.Sprintf("failed to get project %d", projectID), err)
+		return false
+	}
+
+	resource := rbac.NewProjectNamespace(project.ProjectID).Resource(rbac.ResourceNotificationPolicy)
+	if !w.SecurityCtx.Can(action, resource) {
+		w.SendForbiddenError(errors.New(w.SecurityCtx.GetUsername()))
+		return false
+	}
+	return true
+}
+
+func (w *NotificationPolicyAPI) validatePolicyExist() bool {
```

Comment:
`validatePolicyExist` This is not safe in concurrent situation.

Reply:
> `validatePolicyExist` This is not safe in concurrent situation.

fixed, add constraint for project_id on table `notification_policy`